src/runtime/cuda/nccl_api.cu

```diff
@@ -0,0 +1,832 @@
+      col += blockDim.x;
+    }
+    in_row += stride;
+  }
+}
+
+
+}  // namespace
+
+/* NCCLUniqueId **************************************************************/
+
+NCCLUniqueId::NCCLUniqueId() :
+  id_() {
+  // this ID is unique to the process, not to each call of this function
+  NCCL_CALL(ncclGetUniqueId(&id_));
+}
+
+ncclUniqueId NCCLUniqueId::Get() const {
+  return id_;
+}
+
+std::string NCCLUniqueId::ToString() const {
+  std::ostringstream oss;
+
+  oss << std::hex;
+
+  for (size_t b = 0; b < NCCL_UNIQUE_ID_BYTES; ++b) {
+    const int num = static_cast<uint8_t>(id_.internal[b]);
+    oss << std::setw(2) << std::setfill('0') << num;
+  }
+
+  std::string result = oss.str();
+  CHECK_EQ(result.length(), NCCL_UNIQUE_ID_BYTES*2) <<
+    "Invalid NCCL ID format: '" << result << "'";
+
+  return result;
+}
+
+void NCCLUniqueId::FromString(
+    const std::string& str) {
+  // must be exactly 256 hex characters
+  CHECK_EQ(str.length(), NCCL_UNIQUE_ID_BYTES * 2) <<
+        "Invalid NCCL ID format: '" << str << "'";
+
+  for (size_t b = 0; b < NCCL_UNIQUE_ID_BYTES; ++b) {
+    id_.internal[b] = std::strtol(str.substr(b*2, 2).c_str(), nullptr, 16);
+  }
+}
+
+template<typename IdType, typename DType>
+std::pair<IdArray, NDArray> SparsePush(
+    NCCLCommunicatorRef comm,
+    IdArray in_idx,
+    NDArray in_value,
+    NDArrayPartitionRef part) {
+  CHECK_EQ(in_idx->shape[0], in_value->shape[0]);
+
+  const auto& ctx = in_idx->ctx;
+  CHECK_EQ(ctx, in_value->ctx);
+  auto device = DeviceAPI::Get(ctx);
+
+  // TODO(dlasalle): Get the stream from the device context.
+  cudaStream_t stream = 0;
+
+  CHECK_EQ(in_idx->ndim, 1);
+
+  const int64_t num_in = in_idx->shape[0];
+  int64_t num_feat = 1;
+  for (int d = 1; d < in_value->ndim; ++d) {
+    num_feat *= in_value->shape[d];
+  }
+
+  const int64_t comm_size = comm->size();
+
+  if (comm_size == 1) {
+    // nothing to do, just return original arrays
+    return std::pair<IdArray, NDArray>(in_idx, in_value);
+  }
+
+  std::pair<IdArray, NDArray> part_perm = part->GeneratePermutation(in_idx);
+  const IdType * const perm = static_cast<const IdType*>(part_perm.first->data);
+  const int64_t * const send_sum =
+      static_cast<const int64_t*>(part_perm.second->data);
+
+  IdType * send_idx = static_cast<IdType*>(device->AllocWorkspace(ctx,
+      num_in*sizeof(IdType)));
+  DType * send_value = static_cast<DType*>(device->AllocWorkspace(ctx,
+      num_in*num_feat*sizeof(DType)));
+
+  // permute the indices and values
+  {
+    const dim3 block(256);
+    const dim3 grid((num_in+block.x-1)/block.x);
+
+    _DualPermKernel<<<grid, block, 0, stream>>>(
+        static_cast<const IdType*>(in_idx->data),
+        static_cast<const DType*>(in_value->data),
+        perm,
+        num_in,
+        num_feat,
+        send_idx,
+        send_value);
+    CUDA_CALL(cudaGetLastError());
+  }
+
+  // compute the prefix sum of the send values
+  int64_t * send_prefix = static_cast<int64_t*>(
```

Comment:
Where to release send_prefix? Or it is released automatically?

Reply:
I switched all workspace allocations to use RAII via https://github.com/dmlc/dgl/pull/2825/files#diff-128b2340cfe395c91243eab36571e6f38af6b4eb0d0c3e01eaaba552995d8828 in 
69daf52, to prevent leaking workspace.